model/riscv_vmem.sail

```diff
@@ -0,0 +1,854 @@
+    (Read(_), _)                 => E_Load_Page_Fault(),
+    (Write(_), PTW_Access())     => E_SAMO_Access_Fault(),
+    (Write(_), _)                => E_SAMO_Page_Fault(),
+    (Execute(), PTW_Access())    => E_Fetch_Access_Fault(),
+    (Execute(), _)               => E_Fetch_Page_Fault()
+  }
+}
+
+// PRIVATE
+union PTW_Result = {
+  PTW_Success: (bits(64), bits(64), bits(64), nat, bool, ext_ptw),
+  PTW_Failure: (PTW_Error, ext_ptw)
+}
+
+// ****************************************************************
+// Page Table Walk (PTW)
+
+// Note: 'pt_walk()' is recursive => needs separate 'val' decls
+
+// PRIVATE
+val pt_walk : (SV_Params,
+               bits(64),                     // virtual addr
+               AccessType(ext_access_type),  // Read/Write/ReadWWrite/Execute
+               Privilege,                    // User/Supervisor/Machine
+               bool,                         // mstatus.MXR
+               bool,                         // do_sum
+               bits(64),                     // PT base addr
+               PTW_Level,                    // tree level for this recursive call
+               bool,                         // global translation,
+               ext_ptw)                      // ext_ptw
+              -> PTW_Result
+
+function pt_walk(sv_params,
+                 va,
+                 ac,
+                 priv,
+                 mxr,
+                 do_sum,
+                 pt_base,
+                 level,
+                 global,
+                 ext_ptw) = {
+  let vpn_j      = vpn_j_of_va(sv_params, va, level);
+  let pte_offset = vpn_j << sv_params.log_pte_size_bytes;
+  let pte_addr   = pt_base + pte_offset;
+
+  // In Sv32, physical addrs are actually 34 bits, not XLEN(=32) bits.
+  // Below, 'pte_phys_addr' is XLEN bits because it's an arg to
+  // 'mem_read_priv()' [riscv_mem.sail] where it's declared as xlenbits.
+  // That def and this use need to be fixed together (TODO)
+  let pte_phys_addr : xlenbits = pte_addr[(sizeof(xlen) - 1) .. 0];
+
+  // Read this-level PTE from mem
+  let mem_result = mem_read_priv(Read(Data),              // AccessType
+                                 Supervisor,              // Privilege
+                                 pte_phys_addr,
+                                 8,                       // atom (8)
+                                 false,                   // aq
+                                 false,                   // rl
+                                 false);                  // res
+
+  match mem_result {
+    MemException(_) => PTW_Failure(PTW_Access(), ext_ptw),
+    MemValue(pte)   => {
+      let pte_lsbs = Mk_PTE_LSBs(pte[7 .. 0]);
+      if pte_is_invalid(pte_lsbs) then
+        PTW_Failure(PTW_Invalid_PTE(), ext_ptw)
+      else {
+        // Non-Leaf PTE
+        let ppns : bits(64) = PPNs_of_PTE(sv_params, pte);
+        let global'         = global | (pte_lsbs.G() == 0b1);
```

Comment:
Maybe `next_global` instead of `global'`. Bit weird syntax if you ask me - I mean someone reading this is going to think "what does `'` mean?" whereas `next_global` is obvious. And for the other variables.

Reply:
In my experience, x, x’, x’’, …. is a common idiom in mathematics and programming to indicate next, next-next, etc.